### ISSUE TITLE
Nested Mapped Type Doesn't Cause Error for Mismatch enum when Single Mapped Type Does

### ISSUE THREAD

**Issue by wilwade** [microsoft/TypeScript#45530]
# Bug Report

### 🔎 Search Terms

nested mapping, generic type with enum matching, mapped type over another mapped type

### 🕗 Version & Regression Information

This is the behavior in every version I tried (all playground versions), and I reviewed the FAQ for entries about nested mapped types.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.5#code/KYOwrgtgBAggQgYSgb1gGinDSC+AoPAFwE8AHYKAcVAB4AVKYAD0NABMBnWRAPigF4UUAG4AuKHQDcUfETIVqIAEz1GLdl3gI+g5HihQA2gAUATgHtypklACWIKAGtgxcwDMqtOjwC641BziHISm9gDmMpJ4OFFy5FAAsgCGpORsMABGAMZ05ooCKPpGDPbcCH6eIPQ8UTEEWeYgwVBhoHTyMOKKNFoAdDA6QmJl-ZF4DU2ELW3ycF20fXCDqMOLYxPNrSDt5AjzVX3aBSvih2N4APQXUAjmpqbAWVMAYkm2ADZgD+ONzYQAFuEAOq2AHJVLANjicFpTI5PKgY5FQx9GAVLY7YBYKBXG53B5PKBuN6fB4AQigACVgABHMC2B6JJKELKAkBhMnIxbombkbG4yjmcxsXpMllssJcxC9criDHyBAYAVCkVi1nhaIEEjxGGQuG5RRKJEGQwlBxaCqG6q1WIbKZbJSYzqVFSo5YifxQQJQABEAEkoEloDAfTJ1r97aBHbN9q7pUtjh6hN7-YHoHBQzhw5NpspMXsXT1pUddEmAuJU0Gbpnzri-SAGvdHi8SV9gD8cwDgaD-rq2HQAO62LLAW5NTFQpkQ9LZA1R41GVE8vOzJXXIFJEBTYkfNui6l0hkUZLi8Kck3cuVRzH866C4Wik-q9lShAy5fR3Zrqgqx-M5+SjgQA)

### 💻 Code

```ts
enum ABC { A, B, C }

type Gen<T extends ABC> = { v: T; }

type Gen2<T extends ABC> = {
  [Property in keyof Gen<T>]: { s: string };
};

type MappedAbcToGen = {
  [T in ABC]: Gen<T>;
};

const genTypeA: Gen<ABC.A> = { v: ABC.A };
const genTypeB: Gen<ABC.B> = { v: ABC.B };
const genTypeC: Gen<ABC.C> = { v: ABC.C };

// Correct Failure
const thingWithMapped: MappedAbcToGen = {
  [ABC.A]: genTypeB, // Correct failure! Require Matching!
  [ABC.B]: genTypeB, // Good. Matching
  [ABC.C]: genTypeC, // Good. Matching
}

type MappedAbcToGen2 = {
  [T in ABC]: Gen2<T>;
};

const gen2TypeA: Gen2<ABC.A> = { v: { s: "I am A" } };
const gen2TypeB: Gen2<ABC.B> = { v: { s: "I am B" } };
const gen2TypeC: Gen2<ABC.C> = { v: { s: "I am C" } };

// Incorrect Failure
const thingWithMappedTwiceConsTyped: MappedAbcToGen2 = {
  [ABC.A]: gen2TypeB, // Want failure. Require Matching!
  [ABC.B]: gen2TypeB, // Good. Matching
  [ABC.C]: gen2TypeC, // Good. Matching
}
```

### 🙁 Actual behavior

It did NOT have an error on the double mapped type:
`[ABC.A]: gen2TypeB` should throw an error as it does for `MappedAbcToGen` with `[ABC.A]: genTypeB`

### 🙂 Expected behavior

Same error as with the single mapped type:

`[ABC.A]: gen2TypeB` should throw an error and require `[ABC.A]: gen2TypeA` so that the T matches as `MappedAbcToGen2` implies and works for `MappedAbcToGen`

Thanks for looking!

**Comment by wilwade** [comment]
I realized that perhaps the lack of an error comes from the type not being different at that point in time. Here is an [adjusted playground](https://www.typescriptlang.org/play?ts=4.3.5#code/KYOwrgtgBAggQgYSgb1gGinDSC+AoPAFwE8AHYKAcVAB4AVKYAD0NABMBnWRAPigF4UUAG4AuKHQDcUHFABkUABR4oKFapHj4CAHQw061QENxHQgCcAliADmB1bIA+ajZu66shqACNTF63bqTi4aYu46CPYaAMZ+VrZR+ACUBCTkVKAATPSMLOxc2nyCyOoA2gAK5gD25OYkUNZQANbAxFUAZhkg9DwAuuKoHHEBMpJ4OGNEZBQAskak5Gww3tF0VdQgAiGlDI3a-V09YxME0VUgZlA2oHTTMOIbNNp6RUJhz-pQJlAAREw-ozwZwuhCuN2mcAetGecFeqHeiB0WB84h+xABJ2Bl2uIFu5AQUO6zwQcLcxIwsV+AC8MZMAPR0qAIKrmczAaKggBiRksABswGygedLoQABYBADqljFcwWwDY4lli2Wq3WoC2JVUpQ+BxxeOAyIZTJZbI5UHaPP5bIAhFAAErAACOYEsbKgc0I0XFtmtZRhuvB5ENjMoVSqbB07qMnu9Nj9iIQAdx00iUCNofDkY9XoC41S0yjcqWKzWG0yGrKu02+0J2ToPGOkyxoJxmX19y62Q+pLCg1RAEkvtAYBiMN8+78frJMcKW1l9ZDO09EbCNW4Jz9B0ZoHBRyihENJ9Oxs2wSA2yna8vdCS172oIfN0OmXvKRupw56Yz+yAzqz2VyloCsAQoglAYqStKopKvKdAAO6WNEwDMhc+oKoWyolmq54VlqOriK2C4YEaEpGCAoIWnywGRg6zqurM0Y5j68YeEmF5BsRIZhhGUYxrmeEJmx+qpum3FZoxsbjEAA)

```ts
enum ABC { A, B, C }

type Gen<T extends ABC> = { v: T; } & (
  {
    v: ABC.A,
    a: string,
  } | {
    v: ABC.B,
    b: string,
  } | {
    v: ABC.C,
    c: string,
  }
)

type Gen2<T extends ABC> = {
  [Property in keyof Gen<T>]: { s: string };
};

type MappedAbcToGen = {
  [T in ABC]: Gen<T>;
};

const genTypeA: Gen<ABC.A> = { v: ABC.A, a: "x" };
const genTypeB: Gen<ABC.B> = { v: ABC.B, b: "y" };
const genTypeC: Gen<ABC.C> = { v: ABC.C, c: "z" };

// Correct Failure
const thingWithMapped: MappedAbcToGen = {
  [ABC.A]: genTypeB, // Correct failure! Require Matching!
  [ABC.B]: genTypeB, // Good. Matching
  [ABC.C]: genTypeC, // Good. Matching
}

type MappedAbcToGen2 = {
  [T in ABC]: Gen2<T>;
};

const gen2TypeA: Gen2<ABC.A> = { v: { s: "I am A" }, a: { s: ""} };
const gen2TypeB: Gen2<ABC.B> = { v: { s: "I am B" }, b: { s: ""} };
const gen2TypeC: Gen2<ABC.C> = { v: { s: "I am C" }, c: { s: ""}  };

// Incorrect Failure
const thingWithMappedTwiceConsTyped: MappedAbcToGen2 = {
  [ABC.A]: gen2TypeB, // Want failure. Require Matching!
  [ABC.B]: gen2TypeB, // Good. Matching
  [ABC.C]: gen2TypeC, // Good. Matching
}
```

**Comment by andrewbranch** [comment]
Slightly reduced with some breadcrumbs added: [playground](https://www.typescriptlang.org/play?ts=4.5.0-dev.20210823#code/KYOwrgtgBAggQgYSgb1gGinKBfAULgFwE8AHYKAcVAB4AVKYADwNABMBnWRAPigF4UUAG4AuKLQDcOKADIoAClxQUS5cLHwEAOhhpVygIZj2BAE4BLEAHM9y7FAA+Kteq7a4ttQCNjZyzdU8AEp8AHpQqABFMHMAYwBrKEsAMwB7MQB+LNxwnIjiMkpQACY6BmY2Tk1eAWQ8l1EoEwtrCXq8XILyKhBS+iYWEA43GucoAG0ABVNUslNiJJAoeOAiVOSikDpuAF1fFqs27Db6gHIDU6gDIahTr0uSGbmCc2BOU2AARxiP1lxY1IgExQKwlWikYAwMQ9UqaHSjVCNKAAIgAkldoDBkRgjCjkTg2gCgQQQWCIXBoSVqHC4AjXCj0QZoHBsVAfHiCWEIqchJcshlCBCoABpfjLVbrTawxDwk7hKAiRSsYCxAA2Bg+UFVwBJuJh1JlMG4bWVao15G1JPZ+ppxtwBjFXjaXjFBjaQA)

```ts
enum ABC { A, B }

type Gen<T extends ABC> = { v: T; } & (
  {
    v: ABC.A,
    a: string,
  } | {
    v: ABC.B,
    b: string,
  }
)

// Quick info: ???
//
// type Gen2<T extends ABC> = {
//    v: string;
// }
//
type Gen2<T extends ABC> = {
  [Property in keyof Gen<T>]: string;
};

// 'a' and 'b' properties required
const gen2TypeA: Gen2<ABC.A> = { v:  "I am A", a: "" };
const gen2TypeB: Gen2<ABC.B> = { v:  "I am B", b: "" };

// 'v' ???
type K = keyof Gen2<ABC.A>;

// :(
declare let a: Gen2<ABC.A>;
declare let b: Gen2<ABC.B>;
a = b;
b = a;
```
